homeassistant/components/nissan_leaf.py

```diff
@@ -0,0 +1,467 @@
+DEFAULT_CHARGING_INTERVAL = timedelta(minutes=15)
+DEFAULT_CLIMATE_INTERVAL = timedelta(minutes=5)
+RESTRICTED_BATTERY = 2
+RESTRICTED_INTERVAL = timedelta(hours=12)
+
+MAX_RESPONSE_ATTEMPTS = 22
+
+PYCARWINGS2_SLEEP = 30
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_REGION): vol.In(CONF_VALID_REGIONS),
+        vol.Optional(CONF_NCONNECT, default=True): cv.boolean,
+        vol.Optional(CONF_INTERVAL, default=DEFAULT_INTERVAL): (
+            vol.All(cv.time_period, vol.Clamp(min=MIN_UPDATE_INTERVAL))),
+        vol.Optional(CONF_CHARGING_INTERVAL,
+                     default=DEFAULT_CHARGING_INTERVAL): (
+                         vol.All(cv.time_period,
+                                 vol.Clamp(
+                                     min=MIN_UPDATE_INTERVAL))),
+        vol.Optional(CONF_CLIMATE_INTERVAL,
+                     default=DEFAULT_CLIMATE_INTERVAL): (
+                         vol.All(cv.time_period,
+                                 vol.Clamp(
+                                     min=MIN_UPDATE_INTERVAL))),
+        vol.Optional(CONF_FORCE_MILES, default=False): cv.boolean
+    })])
+}, extra=vol.ALLOW_EXTRA)
+
+LEAF_COMPONENTS = [
+    'sensor', 'switch', 'binary_sensor', 'device_tracker'
+]
+
+SIGNAL_UPDATE_LEAF = 'nissan_leaf_update'
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Set-up the Nissan Leaf component."""
+    import pycarwings2
+
+    async def async_setup_leaf(car_config):
+        """Set-up a car."""
+        _LOGGER.debug("Logging into You+Nissan...")
+
+        username = car_config[CONF_USERNAME]
+        password = car_config[CONF_PASSWORD]
+        region = car_config[CONF_REGION]
+        leaf = None
+
+        async def leaf_login():
+            nonlocal leaf
+            sess = pycarwings2.Session(username, password, region)
+            leaf = sess.get_leaf()
+
+        try:
+            # this might need to be made async (somehow) causes
+            # homeassistant to be slow to start
+            await hass.async_add_job(leaf_login)
+        except(RuntimeError, urllib.error.HTTPError):
+            _LOGGER.error(
+                "Unable to connect to Nissan Connect with "
+                "username and password")
+            return False
+        except KeyError:
+            _LOGGER.error(
+                "Unable to fetch car details..."
+                " do you actually have a Leaf connected to your account?")
+            return False
+        except pycarwings2.CarwingsError:
+            _LOGGER.error(
+                "An unknown error occurred while connecting to Nissan: %s",
+                sys.exc_info()[0])
+            return False
+
+        _LOGGER.info("Successfully logged in and fetched Leaf info")
+        _LOGGER.info(
+            "WARNING: This may poll your Leaf too often, and drain the 12V"
```

Comment:
it's says WARNING.... maybe should be _LOGGER.warning

Reply:
This changed in commit https://github.com/home-assistant/home-assistant/pull/19786/commits/72924b812409f026ae6a7ec694f44c0cd9e1c14a